mooncake-integration/store/store_py.cpp

```diff
@@ -725,6 +725,158 @@ int DistributedObjectStore::get_into(const std::string &key, void *buffer,
     return static_cast<int>(total_size);
 }
 
+std::vector<int> DistributedObjectStore::batch_put_from(
+    const std::vector<std::string> &keys, const std::vector<void *> &buffers,
+    const std::vector<size_t> &sizes) {
+    if (!client_) {
+        LOG(ERROR) << "Client is not initialized";
+        return std::vector<int>(keys.size(), -1);
+    }
+
+    if (keys.size() != buffers.size() || keys.size() != sizes.size()) {
+        LOG(ERROR) << "Mismatched sizes for keys, buffers, and sizes";
+        return std::vector<int>(keys.size(), -1);
+    }
+
+    std::unordered_map<std::string, std::vector<mooncake::Slice>> all_slices;
+    ReplicateConfig config;
+    config.replica_num = 1;  // Make configurable
+    config.preferred_segment = this->local_hostname;
+
+    for (size_t i = 0; i < keys.size(); ++i) {
+        const auto &key = keys[i];
+        void *buffer = buffers[i];
+        size_t size = sizes[i];
+
+        if (size == 0) {
+            LOG(WARNING) << "Attempting to put empty data for key: " << key;
+            continue;
+        }
+
+        std::vector<mooncake::Slice> key_slices;
+        uint64_t offset = 0;
+        while (offset < size) {
+            auto chunk_size = std::min(size - offset, kMaxSliceSize);
+            void *chunk_ptr = static_cast<char *>(buffer) + offset;
+            key_slices.emplace_back(Slice{chunk_ptr, chunk_size});
+            offset += chunk_size;
+        }
+        all_slices[key] = key_slices;
+    }
+
+    ErrorCode batch_put_err = client_->BatchPut(keys, all_slices, config);
+
+    std::vector<int> results(keys.size());
+    if (batch_put_err != ErrorCode::OK) {
+        LOG(ERROR) << "BatchPut failed with error: " << toString(batch_put_err);
+        std::fill(results.begin(), results.end(), toInt(batch_put_err));
+    } else {
+        std::fill(results.begin(), results.end(), 0);
+    }
+
+    return results;
+}
+
+std::vector<int> DistributedObjectStore::batch_get_into(
+    const std::vector<std::string> &keys, const std::vector<void *> &buffers,
+    const std::vector<size_t> &sizes) {
+    auto start_time = std::chrono::steady_clock::now();
+    if (!client_) {
+        LOG(ERROR) << "Client is not initialized";
+        return std::vector<int>(keys.size(), -1);
+    }
+
+    if (keys.size() != buffers.size() || keys.size() != sizes.size()) {
+        LOG(ERROR) << "Mismatched sizes for keys, buffers, and sizes";
+        return std::vector<int>(keys.size(), -1);
+    }
+
+    std::vector<int> results(keys.size());
+    mooncake::Client::BatchObjectInfo
+        object_infos;  // This is BatchGetReplicaListResponse
+
+    // Step 1: Batch query object info
+    ErrorCode batch_query_err = client_->BatchQuery(keys, object_infos);
```

Comment:
It seems like in step 2 and 3, even if there are some non-existent objects, this function will still get the value for the rest of keys. However, in step 1, this function will abort if any of the keys do not exist. Is this intentional or not?

Reply:
We could add an `early_return` flag in the future to control this behavior.